### PR TITLE
[BEAM-10937] Tour of Beam: Reading and writing data notebook

### PR DESCRIPTION
**Pull request by davidcavazos**
This is the second introductory notebook on how to read and write data.

It covers:
* Basic I/O concepts
* `ReadFromText` and `WriteToText` transforms
* Creating new sources
  * Creating data from generators
  * Reading data from SQLite
* Creating new sinks
  * Writing fixed-sized batches
  * Writing windows of elements

R: @aaltay
R: @rosetn

Staged:
* New notebook: https://colab.research.google.com/github/davidcavazos/beam/blob/tour-of-beam-reading-writing-data/examples/notebooks/tour-of-beam/reading-and-writing-data.ipynb
* Updated "What's next" section: https://colab.research.google.com/github/davidcavazos/beam/blob/tour-of-beam-reading-writing-data/examples/notebooks/tour-of-beam/getting-started.ipynb
* Added new entry: http://apache-beam-website-pull-requests.storage.googleapis.com/14045/get-started/tour-of-beam/index.html

> Note: I tried reading from public Cloud Storage data and public BigQuery data, but they all required authentication, so I decided to not include them.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by davidcavazos** [comment]
Will need to be added to the landing page #13747 as well, whichever merges first.

**Comment by aaltay** [comment]
I glanced at this. It looks good. Some high level comments:
- I am not sure the info boxes related to disk speeds are relevant. Read/Write might be mostly happening to a service and very unlikely to use in memory sources.
- The parts that refer to creating a sink transform and creating a source transform might cause confusion. Especially the source one, since source has a specific meaning.

I would like to load balance these reviews across the team. For this one I will nominate @emilymye in addition to @rosetn.

Feel free to ping me once this is ready to merge after both reviews.

**Comment by davidcavazos** [comment]
Got it, I removed the info boxes about the disk speeds. I'm also renaming "Source" to "input transform" and "Sink" to "output transform" since that might be more accurate terms.

**Review by emilymye — COMMENTED** [review]
Will defer to @rosetn for rest of review :)

File: examples/notebooks/tour-of-beam/reading-and-writing-data.ipynb
Comment (davidcavazos) [reply examples/notebooks/tour-of-beam/reading-and-writing-data.ipynb:170]: That's a great suggestion, I'll be changing that. Thanks!

File: examples/notebooks/tour-of-beam/reading-and-writing-data.ipynb
Comment (davidcavazos) [reply examples/notebooks/tour-of-beam/reading-and-writing-data.ipynb]: Sure, that makes sense. Thanks!

File: examples/notebooks/tour-of-beam/reading-and-writing-data.ipynb
Comment (rosetn) [inline examples/notebooks/tour-of-beam/reading-and-writing-data.ipynb:411]: I think you're missing a "to" in this sentence

Comment (rosetn) [inline examples/notebooks/tour-of-beam/reading-and-writing-data.ipynb:411]: Replace `dict`s with "dictionary objects", "`dict` objects", or "Python dictionaries"

File: examples/notebooks/tour-of-beam/reading-and-writing-data.ipynb
Comment (rosetn) [inline examples/notebooks/tour-of-beam/reading-and-writing-data.ipynb:6]: General note about this notebook: consider using text formatting more sparingly. 

* If you're using italics to define terms, don't italicize words that you're not trying to define or explain. I'm actually a little confused about why some terms are italicized; it could be possible that you're trying to make a new user aware of important terms. However, pulling these terms out in a bulleted list to highlight them or naming your headers in a way to direct attention to the term might be a better choice.
* On a similar note with bolding terms, I'm not sure why Source and Sink are bolded. 
* Here are some general guidelines: https://developers.google.com/style/text-formatting. We don't need to follow them strictly, but we should be deliberate in our choices.

File: website/www/site/content/en/get-started/tour-of-beam.md
Comment (rosetn) [inline website/www/site/content/en/get-started/tour-of-beam.md:39]: I'd replace "here" with "In this notebook"

File: website/www/site/content/en/get-started/tour-of-beam.md
Comment (rosetn) [inline website/www/site/content/en/get-started/tour-of-beam.md:33]: We learn about data pipelines, PCollections, and PTransforms, as well as some basic transforms like `Map`, `FlatMap`, `Filter`, `Combine`, and `GroupByKey`.

Or keep them in italics.

**Review by rosetn — CHANGES_REQUESTED** [review]

File: examples/notebooks/tour-of-beam/reading-and-writing-data.ipynb
Comment (davidcavazos) [reply examples/notebooks/tour-of-beam/reading-and-writing-data.ipynb]: Thanks, added.

Comment (davidcavazos) [reply examples/notebooks/tour-of-beam/reading-and-writing-data.ipynb]: Replaced with "Python dictionaries".

File: website/www/site/content/en/get-started/tour-of-beam.md
Comment (davidcavazos) [reply website/www/site/content/en/get-started/tour-of-beam.md]: Removed italics.

File: website/www/site/content/en/get-started/tour-of-beam.md
Comment (davidcavazos) [reply website/www/site/content/en/get-started/tour-of-beam.md]: Done

**Review by davidcavazos — COMMENTED** [review]
Thanks for the comments, here's the updated version.

**Comment by davidcavazos** [comment]
Hi @emilymye, can you take a look at this whenever you have a chance? Thank you!

**Comment by davidcavazos** [comment]
Hi @rosetn, I've addressed your review comments, please let me know what you think.

**Review by rosetn — APPROVED** [review]
Look great, David!

**Comment by davidcavazos** [comment]
Friendly ping @emilymye :)

**Comment by emilymye** [comment]
LGTM! @aaltay will have to actually approve and submit because I don't have committer access.

**Review by emilymye — APPROVED** [review]

**Review by aaltay — APPROVED** [review]

**Comment by aaltay** [comment]
Thank you! And thank you for the reviews!

**Pull request merged**
